### PR TITLE
[ADDED] CleanupPublisher method for removing internal JetStream subscription

### PR DESCRIPTION
**Pull request by piotrpio**
Signed-off-by: Piotr Piotrowski <piotr@synadia.com>

### CODE REVIEW

File: js.go [hunk 1]
```diff
@@ -58,6 +58,19 @@ type JetStream interface {
 	// PublishAsyncComplete returns a channel that will be closed when all outstanding messages are ack'd.
 	PublishAsyncComplete() <-chan struct{}
 
+	// CleanupPublisher will cleanup the publishing side of JetStreamContext.
+	//
+	// This will unsubscribe from the internal reply subject if needed.
+	// All pending async publishes will fail with ErrJetStreamContextClosed.
```
Comment (wallyqs) [inline js.go:64]:
```suggestion
	// All pending async publishes will fail with ErrJetStreamPublisherClosed.
```

File: jetstream/jetstream.go [hunk 2]
```diff
@@ -1032,6 +1045,41 @@ func wrapContextWithoutDeadline(ctx context.Context) (context.Context, context.C
...
+	errCb := js.publisher.aecb
+	for id, paf := range js.publisher.acks {
+		paf.err = ErrJetStreamPublisherClosed
+		if paf.errCh != nil {
+			paf.errCh <- paf.err
+		}
+		if errCb != nil {
+			// clear reply subject so that new one is created on republish
+			js.publisher.Unlock()
+			errCb(js, paf.msg, ErrJetStreamPublisherClosed)
```
Comment (Jarema) [inline jetstream/jetstream.go:1071]: is it safe to unlock and lock again?

File: js.go [hunk 3]
```diff
@@ -734,6 +747,41 @@ func (js *js) resetPendingAcksOnReconnect() {
...
+	errCb := js.opts.aecb
+	for id, paf := range js.pafs {
+		paf.err = ErrJetStreamPublisherClosed
+		if paf.errCh != nil {
+			paf.errCh <- paf.err
+		}
+		if errCb != nil {
+			// clear reply subject so that new one is created on republish
+			js.mu.Unlock()
+			errCb(js, paf.msg, ErrJetStreamPublisherClosed)
```
Comment (Jarema) [inline js.go:773]: same here.

**Review by Jarema — CHANGES_REQUESTED** [review]
Looks ok.  Some questions added.

**Review by Jarema — APPROVED** [review]
LGTM!

**Pull request merged**